pkg/controllers/nodeclaim/garbagecollection/nic_garbagecollection.go

```diff
@@ -0,0 +1,149 @@
+/*
+Portions Copyright (c) Microsoft Corporation.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package garbagecollection
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/patrickmn/go-cache"
+	"github.com/samber/lo"
+	"knative.dev/pkg/logging"
+
+	"github.com/awslabs/operatorpkg/singleton"
+	"k8s.io/client-go/util/workqueue"
+	controllerruntime "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	karpv1 "sigs.k8s.io/karpenter/pkg/apis/v1"
+	"sigs.k8s.io/karpenter/pkg/operator/injection"
+
+	sdkerrors "github.com/Azure/azure-sdk-for-go-extensions/pkg/errors"
+	"github.com/Azure/karpenter-provider-azure/pkg/providers/instance"
+)
+
+const (
+	NICGCControllerName    = "networkinterface.garbagecollection"
+	NicReservationDuration = time.Second * 180
+	// We set this interval at 5 minutes, as thats how often our NRP limits are reset.
+	// See: https://learn.microsoft.com/en-us/azure/azure-resource-manager/management/request-limits-and-throttling#network-throttling
+	NicGarbageCollectionInterval = time.Minute * 5
```

Comment:
So this caches (with eviction) for 3 min, but the next attempt is in 5 min? Wouldn't it find the cache always empty (Have you observed a cache hit for "failed-to-delete" reason during testing?) If the next try is going to be after 5 min, maybe "failed-to-delete" state is not needed at all?

Reply:
no you are correct, I removed the sdkerrors 3 minute retry. I am being conservative with retries as the price of not being able to bootstrap outweighs latent deletion attempts. 

We don't need the IsReserved check at all or the expiration cache after further reflection.